### PR TITLE
adjust width for CJK characters

### LINKED ISSUE #11859: Block cursor width is not correct with CJK characters
- VSCode Version: 1.5.0
- OS Version: OSX

Originally posted in https://github.com/VSCodeVim/Vim/issues/751 but later we found it's a Code issue. 

Repro steps
1. Set `"editor.cursorStyle": "block"`
2. Move cursor to any CJK characters

screen shot 2016-09-11 at 3 26 23 pm

### PR DESCRIPTION
**Pull request by rebornix**
Fix #11859 . Read character from view to get the correct width of a character.

### CODE REVIEW

File: src/vs/editor/browser/viewParts/viewCursors/viewCursor.ts [hunk 1]
```diff
@@ -128,6 +129,17 @@ export class ViewCursor {
 			this._positionTop = visibleRange.top;
 			this._positionLeft = visibleRange.left;
 			this._isInViewport = true;
+
+			let visibleRangeForCharacter = ctx.linesVisibleRangesForRange({
```
Comment (alexandrudima) [inline src/vs/editor/browser/viewParts/viewCursors/viewCursor.ts]: @rebornix This extra call should be done only if the cursor style is a block cursor

File: src/vs/editor/browser/viewParts/viewCursors/viewCursor.ts [hunk 2]
```diff
@@ -156,6 +168,15 @@ export class ViewCursor {
 			this._domNode.setLineHeight(this._lineHeight);
 			this._domNode.setHeight(this._lineHeight);
 
+			let desiredWidth = '1ch';
```
Comment (alexandrudima) [inline src/vs/editor/browser/viewParts/viewCursors/viewCursor.ts]: @rebornix Computing the desiredWidth and setting it needs to happen only if the cursor style is set to block.

**Review by alexdima — CHANGES_REQUESTED** [review]
Looks ok, but the new code should run only if the cursor is configured to be rendered as a block. Currently, this makes line cursors (the default we ship with) render as a block cursor

**Comment by rebornix** [comment]
Comments addressed. Only update cursor width if its style is block or underline.

@alexandrudima we had issues with underline cursor style as well, which should be taken care of the same way as block cursor. Besides, this fix makes tab more explicit as below

screen shot 2016-11-27 at 1 56 01 pm
screen shot 2016-11-27 at 1 56 15 pm

**Review by alexdima — APPROVED** [review]

**Comment by alexandrudima** [comment]
:+1:

**Pull request merged**